addon-api/content-script/blocks.js

```diff
@@ -0,0 +1,206 @@
+  let oldUpdateColor = BlockSvg.prototype.updateColour;
+  BlockSvg.prototype.updateColour = function (...a) {
+    if (this.procCode_) {
+      let p = this.procCode_;
+      let block = customBlocks.find((e) => e.id == p.trim());
+      if (block) {
+        this.colour_ = block.color;
+        this.colourSecondary_ = block.secondaryColor;
+        this.colourTertiary_ = block.tertiaryColor;
+        let updateChildColors = function updateChildColors() {
+          this.childBlocks_.forEach(
+            ((e) => {
+              e.setColour(e.getColour(), e.getColourSecondary(), this.getColourTertiary());
+            }).bind(this)
+          );
+        }.bind(this);
+        updateChildColors();
+        const oldPush = this.childBlocks_.constructor.prototype.push.bind(this.childBlocks_);
+        this.childBlocks_.push = function (...a) {
+          updateChildColors();
+          return oldPush(...a);
+        };
+      }
+    }
+    return oldUpdateColor.call(this, ...a);
+  };
+
+  const flyout = workspace.getFlyout();
+  if (!flyout) throw new Error("expected flyout");
+  if (!vm) throw new Error("expected vm");
+  // Each time a new workspace is made, these callbacks are reset, so re-register whenever a flyout is shown.
+  // https://github.com/LLK/scratch-blocks/blob/61f02e4cac0f963abd93013842fe536ef24a0e98/core/flyout_base.js#L469
+  const originalShow = flyout.constructor.prototype.show;
+  flyout.constructor.prototype.show = function (xml) {
+    this.workspace_.registerToolboxCategoryCallback("SABLOCKS", function (e) {
+      return [
+        ...new DOMParser()
+          .parseFromString(
+            `<top>` +
+              (customBlocks
+                .filter((e) => !e.hide)
+                .map((e) => {
+                  try {
+                    return `<block type="procedures_call" gap="16"><mutation generateshadows="true" proccode="${xesc(
+                      e.id
+                    )}" argumentids="${xesc(JSON.stringify(e.args.map((e, i) => "arg" + i)))}" argumentnames="${xesc(
+                      JSON.stringify(e.args)
+                    )}" argumentdefaults="${xesc(
+                      JSON.stringify(e.args.map((e) => ""))
+                    )}" warp="false"></mutation></block>`;
+                  } catch (e) {
+                    console.error("Issue making block from", e);
+                    return "";
+                  }
+                })
+                .join("") ||
+                `<label text="${xesc(
+                  scratchAddons.l10n.get("noAddedBlocks", null, "No addons have added blocks.")
+                )}" showStatusButton="null" />`) +
+              `</top>`,
+            "text/xml"
+          )
+          .querySelectorAll("block, label"),
+      ];
+    });
+    return originalShow.call(this, xml);
+  };
+
+  // We use Scratch's extension category mechanism to replace the data category with our own.
```

Comment:
This code was taken from `data-category-tweaks-v2` right?

Reply:
Taken from a11y which took from data category tweaks.